### PR TITLE
make ls work like dir on windows

### PR DESCRIPTION
**Pull request by richardmarklund**
# Description

Related issue: https://github.com/nushell/nushell/issues/6857

This fix makes the ls command work like dir on windows. This hides more files like symfiles and other dirs that the dir command does not show.

Could not create any new tests since we have a strange way to create hidden files (by creating dotfiles) and could not find any way to create symlinks. 

# Tests + Formatting

Make sure you've done the following, if applicable:

- Add tests that cover your changes (either in the command examples, the crate/tests folder, or in the /tests folder)
  - Try to think about corner cases and various ways how your changes could break. Cover those in the tests

Make sure you've run and fixed any issues with these commands:

- `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- `cargo clippy --workspace --features=extra -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect` to check that you're using the standard code style
- `cargo test --workspace --features=extra` to check that all tests pass

# User-Facing Changes

If you're making changes that will affect the user experience of Nushell (ex: adding/removing a command, changing an input/output type, adding a new flag):

- Get another regular contributor to review the PR before merging
- Make sure that there is an entry in the documentation (https://github.com/nushell/nushell.github.io) for the feature, and update it if necessary

### CODE REVIEW

File: crates/nu-command/src/filesystem/ls.rs [hunk 1]
```diff
@@ -352,6 +352,7 @@ fn should_skip_dir(dir: impl AsRef<Path>) -> bool {
 }
 
 fn is_hidden_dir(metadata_attributes: u32) -> bool {
+    #[cfg(windows)]
```
Comment (fdncred) [inline crates/nu-command/src/filesystem/ls.rs:355]: I'm guessing that this should be moved up so that `is_hidden_dir` is only available on windows.

**Comment by rgwood** [comment]
Does this change _always_ hide the directories? Or are they visible with the `-a`/`--all` flag?

**Comment by richardmarklund** [comment]
@rgwood
All is visible with -a/--all
No change there.

File: crates/nu-command/src/filesystem/ls.rs (hunk 1 above)
Comment (richardmarklund) [reply crates/nu-command/src/filesystem/ls.rs:355]: Thanks!

**Comment by rgwood** [comment]
@richardmarklund Looks like this has broken tests on Linux. Are you able to look into that or do you need help?

**Comment by richardmarklund** [comment]
@rgwood 
Yea, noticed this morning. I'll look into it

**Comment by dandavison** [comment]
I see this is Windows-only, but what's the motivation for hiding symlinks? (other than emulating a Windows command)

**Comment by rgwood** [comment]
Windows does allow symlinks (but they're a little harder to create than on *nix).

I think the logic needs to be more complex than "is it a symlink?" because cmd.exe and PowerShell _do_ show some symlinks. However, Nu shows a few symlinks that don't show up in cmd.exe and PowerShell:

image

I'm not sure what exactly differentiates those symlinks from the ones shown by cmd.exe and PowerShell.

#### symlinks in cmd.exe dir

image

#### symlinks in PowerShell ls

image

**Comment by ChrisDenton** [comment]
They have the system attribute (as well as having the hidden attribute). In powershell try `ls -System -Hidden`.

**Comment by richardmarklund** [comment]
@rgwood  @ChrisDenton 
Hello and thanks for your comments. That's the way I went from the start. The problem is that I still find differences when I run `ls` in powershell and `ls` in the nushell. 
Powershell:
 
image

nushell:
image


So here I'm filtering out everything that's either system or hidden according to:
https://docs.microsoft.com/en-us/windows/win32/fileio/file-attribute-constants

But if I also remove the symlinks it seems like its correct:
image

**Comment by richardmarklund** [comment]
I also have a question:
The test that is failing now seems a little strange. So we have a test that is passing and works fine named: `lists_all_hidden_files_when_glob_contains_dot` that creates the following structure and returns 3 which is correct according to the test.
root1.txt
root2.txt
.dotfile1
dir_a
     yehuda.10.txt
     jonathan.10.txt
     .dotfile2
     dir_b
           andres.10.txt
           chicken_not_to_be_picked_up.100.txt
           .dotfile3
           
command: `ls **/.*  | length`


The next test: lists_all_hidden_files_when_glob_does_not_contain_dot
Has the same file structure but with this command:
`ls **/*  | length`

The test wants that to return 5, which I find a bit strange. My returns 7 which also feels strange. My intuition wanted it to be 11.  But I guess the command is run inside `dir_a`?

Anyways, how do you want this to work? My thought was that if you specify a glob pattern then you want all the results regardless of if the command is with a dot or not.

We could either go the route of if you are specifying a glob pattern and then give the full result or you'll have to add `-a/-all` to get the full result regardless of the glob pattern.

What's your thought?

**Comment by ChrisDenton** [comment]
Sorry, I'm not very familiar with the code. How are you getting the attributes? The relevant attributes are on the symlink. If you have a `DirEntry` you can use the [`DirEntry::metadata`](https://doc.rust-lang.org/std/fs/struct.DirEntry.html#method.metadata) method but if you just have a path you'll need to use [`Path::symlink_metadata`](https://doc.rust-lang.org/std/path/struct.Path.html#method.symlink_metadata) (or [`GetFileAttributesW`](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getfileattributesw) but that's very Windows specific).

**Comment by rgwood** [comment]
> But if I also remove the symlinks it seems like its correct:

@richardmarklund See my examples [here](https://github.com/nushell/nushell/pull/6990#issuecomment-1309051843); cmd.exe and pwsh do not hide all symlinks by default.

**Comment by deoradh** [comment]
Alternatively, why not just create a switch to opt into os-specific behavior (ie, `ls --os=windows`)?  That, then, could be aliased as desired without inducing a breaking change or affecting the portability of any scripts.  It also allows that behavior to be accessed outside of any specific OS install.

**Comment by rgwood** [comment]
@richardmarklund Thanks for your work on this :). It seems like this PR has stalled so I'm going to close it for now, but if you want to pick it up again we can always reopen it.

**Pull request closed without merging**